// permissionless/_cjs/actions/smartAccount/deployContract.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.deployContract = void 0;
const utils_1 = require("viem/utils");
const utils_2 = require("../../utils/index.js");
const signUserOperationHashWithECDSA_1 = require("../../utils/signUserOperationHashWithECDSA.js");
const waitForUserOperationReceipt_1 = require("../bundler/waitForUserOperationReceipt.js");
const sendUserOperation_1 = require("./sendUserOperation.js");
async function deployContract(client, args) {
    const { abi, args: constructorArgs, bytecode, middleware, ...request } = args;
    const { account: account_ = client.account } = request;
    if (!account_) {
        throw new signUserOperationHashWithECDSA_1.AccountOrClientNotFoundError({
            docsPath: "/docs/actions/wallet/sendTransaction"
        });
    }
    const account = (0, utils_2.parseAccount)(account_);
    const userOpHash = await (0, utils_1.getAction)(client, (sendUserOperation_1.sendUserOperation), "sendUserOperation")({
        userOperation: {
            sender: account.address,
            maxFeePerGas: request.maxFeePerGas || BigInt(0),
            maxPriorityFeePerGas: request.maxPriorityFeePerGas || BigInt(0),
            callData: await account.encodeDeployCallData({
                abi,
                bytecode,
                args: constructorArgs
            })
        },
        account: account,
        middleware
    });
    const userOperationReceipt = await (0, utils_1.getAction)(client, waitForUserOperationReceipt_1.waitForUserOperationReceipt, "waitForUserOperationReceipt")({
        hash: userOpHash
    });
    return userOperationReceipt?.receipt.transactionHash;
}
exports.deployContract = deployContract;
