// permissionless/_cjs/actions/smartAccount/sendTransactions.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.sendTransactions = void 0;
const utils_1 = require("viem/utils");
const utils_2 = require("../../utils/index.js");
const waitForUserOperationReceipt_1 = require("../bundler/waitForUserOperationReceipt.js");
const sendUserOperation_1 = require("./sendUserOperation.js");
async function sendTransactions(client, args) {
    const { account: account_ = client.account, transactions, middleware, maxFeePerGas, maxPriorityFeePerGas, nonce } = args;
    if (!account_) {
        throw new utils_2.AccountOrClientNotFoundError({
            docsPath: "/docs/actions/wallet/sendTransaction"
        });
    }
    const account = (0, utils_2.parseAccount)(account_);
    if (account.type !== "local") {
        throw new Error("RPC account type not supported");
    }
    const callData = await account.encodeCallData(transactions.map(({ to, value, data }) => {
        if (!to)
            throw new Error("Missing to address");
        return {
            to,
            value: value || BigInt(0),
            data: data || "0x"
        };
    }));
    const userOpHash = await (0, utils_1.getAction)(client, (sendUserOperation_1.sendUserOperation), "sendUserOperation")({
        userOperation: {
            sender: account.address,
            maxFeePerGas: maxFeePerGas || BigInt(0),
            maxPriorityFeePerGas: maxPriorityFeePerGas || BigInt(0),
            callData: callData,
            nonce: nonce
        },
        account: account,
        middleware
    });
    const userOperationReceipt = await (0, utils_1.getAction)(client, waitForUserOperationReceipt_1.waitForUserOperationReceipt, "waitForUserOperationReceipt")({
        hash: userOpHash
    });
    return userOperationReceipt?.receipt.transactionHash;
}
exports.sendTransactions = sendTransactions;
