// permissionless/_esm/experimental/eip7677/actions/getPaymasterStubData.js
import { ChainNotFoundError, toHex } from "viem";
import { deepHexlify, getEntryPointVersion } from "../../../utils/index.js";
export async function getPaymasterStubData(client, { userOperation, entryPoint, context, chain }) {
    const chainId = chain?.id ?? client.chain?.id;
    if (!chainId) {
        throw new ChainNotFoundError();
    }
    const params = context
        ? [
            deepHexlify(userOperation),
            entryPoint,
            toHex(chainId),
            context
        ]
        : [
            deepHexlify(userOperation),
            entryPoint,
            toHex(chainId)
        ];
    const response = await client.request({
        method: "pm_getPaymasterStubData",
        params
    });
    const entryPointVersion = getEntryPointVersion(entryPoint);
    if (entryPointVersion === "v0.6") {
        const responseV06 = response;
        return {
            paymasterAndData: responseV06.paymasterAndData
        };
    }
    const responseV07 = response;
    return {
        paymaster: responseV07.paymaster,
        paymasterData: responseV07.paymasterData,
        paymasterVerificationGasLimit: responseV07.paymasterVerificationGasLimit
            ? BigInt(responseV07.paymasterVerificationGasLimit)
            : undefined,
        paymasterPostOpGasLimit: responseV07.paymasterPostOpGasLimit
            ? BigInt(responseV07.paymasterPostOpGasLimit)
            : undefined
    };
}
